platform/ontology/api.go

```diff
@@ -45,99 +47,165 @@ func (p *Platform) GetTokenTxsByAddress(address string, token string) (blockatla
 	}
 	var txs []blockatlas.Tx
 	for _, srcTx := range txPage.Result.TxnList {
-		tx, ok := Normalize(&srcTx, token)
+		tx, ok := Normalize(srcTx, token)
 		if !ok {
 			continue
 		}
-		txs = append(txs, tx)
+		txs = append(txs, *tx)
 	}
 
 	return txs, nil
 }
 
+func Normalize(srcTx Tx, assetName string) (tx *blockatlas.Tx, ok bool) {
+	if len(srcTx.TransferList) < 1 {
+		return tx, false
+	}
+	transfer := srcTx.TransferList[0]
+	fee := numbers.DecimalExp(srcTx.Fee, ONGDecimalsAmount)
+	tx = &blockatlas.Tx{
+		ID:     srcTx.TxnHash,
+		Coin:   coin.ONT,
+		Date:   srcTx.TxnTime,
+		Block:  srcTx.Height,
+		Status: calculateTxStatus(int(srcTx.ConfirmFlag)),
+		Fee:    blockatlas.Amount(fee),
+	}
+
+	switch assetName {
+	case ONTAssetName:
+		normalizeONT(tx, transfer)
+	case ONGAssetName:
+		normalizeONG(tx, transfer)
+	default: // unsupported asset
+		return tx, false
+	}
+	return tx, true
+}
+
 func (p *Platform) CurrentBlockNumber() (int64, error) {
 	block, err := p.client.CurrentBlockNumber()
 	if err != nil {
 		logger.Error("CurrentBlockNumber", logger.Params{"platform": p.Coin().Symbol, "details": err.Error()})
 		return 0, err
 	}
 	var height int64
-	if block.Error != 0 {
-		err = errors.E("explorer error")
-	}
 	if len(block.Result) > 0 {
 		height = (int64)(block.Result[0].Height)
 	}
 	return height, nil
 }
 
-func (p *Platform) GetBlockByNumber(num int64) (*blockatlas.Block, error) {
-	response, err := p.client.GetBlockByNumber(num)
+func (p *Platform) GetBlockByNumber(num int64) (block *blockatlas.Block, err error) {
+	blockOnt, err := p.client.GetBlockByNumber(num)
+	if err != nil {
+		return nil, err
+	}
+	txsRaw, err := p.getTxDetails(blockOnt.Result.TxnList)
 	if err != nil {
-		logger.Error("GetBlockByNumber", logger.Params{"platform": p.Coin().Symbol, "details": err.Error()})
 		return nil, err
 	}
-	var (
-		block blockatlas.Block
-		txs   []blockatlas.Tx
-	)
-	if response.Error == 0 {
-		block.ID = response.Result.Hash
-		block.Number = int64(response.Result.Height)
-		for _, txn := range response.Result.TxnList {
-			tx := new(blockatlas.Tx)
-			tx.ID = txn.TxnHash
-			tx.Block = uint64(txn.Height)
-			if txn.ConfirmFlag == 1 {
-				tx.Status = blockatlas.StatusCompleted
+
+	block = normalizeBlock(*blockOnt, txsRaw)
+	return
+}
+
+func (p *Platform) getTxDetails(txnList []Tx) ([]TxV2, error) {
+	var wg sync.WaitGroup
+	txsChan := make(chan TxV2, len(txnList))
+	wg.Add(len(txnList))
+	for _, blockTxRaw := range txnList {
+		go func(blockTxRaw Tx, wg *sync.WaitGroup) {
+			defer wg.Done()
+			txRaw, err := p.client.GetTxDetailsByHash(blockTxRaw.TxnHash)
+			if err == nil {
+				txsChan <- txRaw
 			}
-			tx.Date = int64(txn.TxnTime)
-			tx.Coin = coin.Ontology().ID
-			txs = append(txs, *tx)
+		}(blockTxRaw, &wg)
+	}
+	wg.Wait()
+	close(txsChan)
```

Comment:
```suggestion
	go func() {
		wg.Wait()
		close(out)
	}()
```

Use a goroutine to close the channel, following the Golang async pattern

Reply:
Got it, however, it does not work at my case, if you can fix it - will be great

Reply:
why? see fan-in, fan-out, is the same example